src/dos/program_setver.cpp

```diff
@@ -0,0 +1,721 @@
+	}
+
+	return stream;
+}
+
+void SETVER::LoadTableFromFile()
+{
+	// Do nothing if no file name specified in the configuration
+	const auto file_path = GetTableFilePath();
+	if (file_path.empty()) {
+		return;
+	}
+
+	// Helper warning lambdas
+
+	bool already_warned_format  = false;
+	bool already_warned_version = false;
+	bool already_warned_name    = false;
+
+	auto warn_file_format = [&]() {
+		if (already_warned_format) {
+			return;
+		}
+		LOG_WARNING("DOS: SETVER - table file seems to be of extended format, ignoring extra data");
+		already_warned_format = true;
+	};
+
+	auto warn_version_parse = [&]() {
+		if (already_warned_version) {
+			return;
+		}
+		LOG_WARNING("DOS: SETVER - problem parsing DOS version");
+		already_warned_version = true;
+	};
+
+	auto warn_file_name = [&]() {
+		if (already_warned_name) {
+			return;
+		}
+		LOG_WARNING("DOS: SETVER - problem parsing file name");
+		already_warned_name = true;
+	};
+
+	// If the file does not exists, save default settings there
+	if (!std_fs::exists(file_path)) {
+		SaveTableToFile();
+		return;
+	}
+
+	// Clear the table - but store previous as a backup
+	const auto backup_table            = setver_table;
+	setver_table.is_global_version_set = false;
+	setver_table.by_file_name.clear();
+	setver_table.by_file_path.clear();
+
+	// Read table from the file
+	std::ifstream file(file_path.c_str());
+	std::string line = {};
+	while (std::getline(file, line)) {
+		// Parse line
+		auto tokens = split(line, '\t');
+		// Skip empty lines
+		if (tokens.empty()) {
+			continue;
+		}
+		// Skip lines with only a single token, we are not currently
+		// using such lines
+		if (tokens.size() == 1) {
+			warn_file_format();
+			continue;
+		}
+		// Only consider first 2 tokens - ignore the remaining ones,
+		// might be needed for some extensions in the future
+		if (tokens.size() > 2) {
+			warn_file_format();
+			tokens.resize(2);
+		}
+		// First token should be a file name - since it is quite likely
+		// the disk is not mounted yet, use permissive preprocesing
+		constexpr bool permissive = true;
+		if (!tokens[0].empty() && !PreprocessName(tokens[0], permissive)) {
+			warn_file_name();
+			continue;
+		}
+		// Second token should be a DOS version
+		FakeVersion version = {};
+		if (!ParseVersion(tokens[1], version)) {
+			warn_version_parse();
+			continue;
+		}
+		// Import row to the table
+		AddToTable(tokens[0], version);
+	}
+	if (file.bad()) {
+		setver_table = backup_table;
+		LOG_WARNING("DOS: SETVER - error reading table file");
+	}
+	file.close();
+}
+
+void SETVER::SaveTableToFile()
+{
+	// Do nothing if no file name specified in the configuration
+	const auto file_path = GetTableFilePath();
+	if (file_path.empty()) {
+		return;
+	}
+
+	// Write table into the file
+	std::ofstream file(file_path.c_str(), std::ios::trunc);
+	if (setver_table.is_global_version_set) {
+		file << "\t" << setver_table.version_global << std::endl;
```

Comment:
Ugh, using `std::endl` will make this file format non-portable; use Unix line endings for all OSes, please.

Reply:
Done.